### ISSUE TITLE
Batched dataloading

### ISSUE THREAD

**Issue by michael-0brien** [patrick-kidger/tinyio#19]
Hello again Patrick! I wanted to share and get your thoughts on some code for loading data in batches using `tinyio`. This is a fun one so this is half show-and-tell (I think other people will find this useful) and half uncertainty in how exactly to implement this.

The idea here is that I have a GPU (as you might expect, JAX) workflow that consumes data whose loading is I/O bound. It is advantageous to load more data at once in batches and formulate a larger GPU computation. Each element of a batch is I/O bound, so it seems like a perfect case for `tinyio`.

TL;DR: a `batch_transform` decorator:

```python
def batch_transform(
    *, batch_size: int, size: int, threads_per_batch: int = DEFAULT_TPB, prefetch=2
):
    """Decorator that consumes a blocking function with signature
    `fn(index, ...)` and returns a co-routine batched over `index`. 
    """

    pool = tinyio.ThreadPool(threads_per_batch)
    semaphore = tinyio.Semaphore(prefetch)
    num_batches = get_num_batches(batch_size, size)

    def batch_coro_impl(fn: Callable[Concatenate[int, P], R]):
        @wraps(coro)
        def batched_coro(
            batch_index: int, *args: P.args, **kwargs: P.kwargs
        ) -> tinyio.Coro[R]:
            with (yield semaphore()):
                assert batch_index < num_batches
                start_index, end_index = (
                    batch_index * batch_size,
                    min((batch_index + 1) * batch_size, size),
                )
                indices = np.asarray(range(start_index, end_index), dtype=int)
                out = yield [
                    pool.run_in_thread(fn, index, *args, **kwargs) for index in indices
                ]
                return list_to_batched(out)  # a function that runs `np.stack` on every array returned by `coro` (using `jax.tree.map`, unsurprisingly)

        return batched_coro

    return batch_coro_impl

def get_num_batches(batch_size: int, size: int) -> int:
    num_batches = size // batch_size
    if size % batch_size != 0:
        num_batches += 1
    return num_batches
```

Usage:

```python

# For the sake of argument, imagine this as a torch-like Dataset,
# but think of `dataset[index]` as an I/O bound blocking function that
# returns an array (or arrays)
dataset = ...

@batch_transform(batch_size=batch_size, size=len(dataset), threads_per_batch=8, prefetch=2)
def load_coro(index, dataset):
    return index, dataset[index]

num_batches = get_num_batches(batch_size, len(dataset))
async_iterator = yield tinyio.as_completed(
      {load_coro(index, dataset) for index in range(num_batches)}
  )
while not async_iterator.done():
    index_batch, data_batch = yield async_iterator.get()
    # Do some work on the GPU
    ...
```

Okay so in practice, I'm not totally confident in how this works; in particular it is a head scratcher to me in how to implement `prefetch`. In the current examples in `tinyio`, `as_completed` runs blocking functions in different threads using `run_in_thread`. This implementation hands `load_coro` directly to `as_completed`, which is already transformed to a co-routine by `batch_transform`.

Therefore, I don't understand exactly what parallelism occurs under the hood using `prefetch` (and I'm a little out of my depth in profiling this), and if it yields any benefits. Is it that the scheduler will attempt to load different batches concurrently, but if there are other blocking functions the in the program there may not be improved performance?

**Comment by michael-0brien** [comment]
Thanks so much for the feedback! Hopefully other people find this useful as well. For completeness, I’ve put the modified version of the code at the bottom here.

> Looking over what you're done, I think it looks essentially correct. `as_completed` definitely doesn't have to be paired with `run_in_thread`.

Things are starting to click of what’s going on under the hood!

> The same rules don't apply to something that is running in a thread though. The event loop needs to at least _start_ the thread (at some point after `yield run_in_thread(...)` has happened), but once started then it will keep running in the background regardless.

Got it!

> Looking at what you have, I think you can improve your implementation here:
> 
> 1. Neatness: a common trick to implement `get_num_batches` is `ceildiv = lambda a, b: (a + b - 1) // b`.

Thanks! 

> 2. Bugfix: note that your threadpool is not allocating that many threads per batch, but is allocating that many threads across all batches.

Aha, thanks for the catch!

> 3. Bigger point: suppose the 'Do some work on the GPU' takes an incredibly long time. Then (a) not all background threads may get a chance to be scheduled (e.g. because they're part of the pool), and (b) whilst all background threads that have started will complete, their `list_to_batched` will not have ran yet.>    Noodling on this point a bit, it becomes clear that you probably need at least two threads: one to do the GPU work, and one to orchestrate the dataloading. If one of these takes a really long blocking time, then you probably don't want the other to block.>

Ok this is interesting and makes sense. It sounds like this can be implemented with a simple `yield pool.run_in_thread(do_gpu_work, …)`, with `max_threads = 1`.

But this evokes the larger question, which others may find helpful for discussion: how does the event loop / threads pair with libraries that have GPU async dispatch (e.g. JAX)?

Naively I would think this makes it so the event loop can proceed, as long as the user doesn’t ask for the result. Though running in a thread I suppose would allow the event loop to proceed even in this case.

> Honestly though, dataloading is not something I'd normally use an event loop / coroutines for. I think this problem is fairly well-modelled with just threads. (Unless maybe you're looking to handle some edge-case stuff like graceful shutdown from SIGINTs, which tinyio will communicate to its threads but default Python will not.)

I suppose I really am just using `tinyio` as a wrapper for launching a bunch of threads, so its niceties like graceful shutdown really do end up being a big reason. But my use case really is quite a bit simpler than the typical ML use case (which is why some other libraries are overkill for me). My “dataloading” really is as simple as loading a bunch of files plus some light CPU work.

The modified decorator:

```python
def batch_transform(
    *, batch_size: int, size: int, threads_per_batch: int = 8, prefetch=2
):
    """Decorator that consumes a blocking function with signature
    `fn(index, ...)` and returns a co-routine batched over `index`. 
    """

    semaphore = tinyio.Semaphore(prefetch)
    num_batches = get_num_batches(batch_size, size)

    def batch_coro_impl(fn: Callable[Concatenate[int, P], R]):
        pool = tinyio.ThreadPool(threads_per_batch)
        @wraps(coro)
        def batched_coro(
            batch_index: int, *args: P.args, **kwargs: P.kwargs
        ) -> tinyio.Coro[R]:
            with (yield semaphore()):
                assert batch_index < num_batches
                start_index, end_index = (
                    batch_index * batch_size,
                    min((batch_index + 1) * batch_size, size),
                )
                indices = np.asarray(range(start_index, end_index), dtype=int)
                out = yield [
                    pool.run_in_thread(fn, index, *args, **kwargs) for index in indices
                ]
                return list_to_batched(out)  # a function that runs `np.stack` on every array returned by `fn`

        return batched_coro

    return batch_coro_impl

def get_num_batches(batch_size: int, size: int) -> int:
    return (size + batch_size - 1) // batch_size
```

Usage:

```python
# Like a torch-like Dataset,
# but think of `dataset[index]` as an I/O bound blocking function that
# returns an array (or arrays)
dataset = ...

@batch_transform(batch_size=batch_size, size=len(dataset), threads_per_batch=8, prefetch=2)
def load_coro(index, dataset):
    return index, dataset[index]

def run_gpu(data):
    “””Heavy GPU work”””
    ….

work_pool = tinyio.ThreadPool(1)
num_batches = get_num_batches(batch_size, len(dataset))
async_iterator = yield tinyio.as_completed(
      {load_coro(index, dataset) for index in range(num_batches)}
  )
while not async_iterator.done():
    index_batch, data_batch = yield async_iterator.get()
    out = yield work_pool.run_in_thread(run_gpu, data_batch)
```
